### ISSUE TITLE
Universal Developer Image

### ISSUE THREAD

**Issue by l0rd** [devfile/api#616]
/area registry

## Which functionality do you think we should add?

Devfile is a specification for containerized development environments. Such environments need to have a minimal set of command line developers tools. 

For that reason we should define an image (the universal developer image) used by the devfile registry containers.

## Describe the solution you'd like

Here is a list of tools required in the universal developer image

**base terminal tools**

| Tool                | ubi8 based image                    |
|---------------------|-------------------------------------|
| `bash`              |`bash`                               |
| `bat`               |`<gh releases>`                      |
| `curl`              |`curl`                               |
| `diff`              |`diffutils`                          |
| `emacs`             |`NOT AVAILABLE (fedora only)`        |
| `fish`              |`NOT AVAILABLE (fedora only)`        |
| `gh`                |`<gh releases>`                      |
| `git`               |`git`                                |
| `ip`                |`iproute`                            |
| `jq`                |`jq`                                 |
| `htop`              |`NOT AVAILABLE (fedora only)`        |
| `less`              |`less`                               |
| `lsof`              |`lsof`                               |
| `man`               |`man`                                |
| `nano`              |`nano`                               |
| `netcat`            |`NOT AVAILABLE`                      |
| `netstat`           |`net-tools`                          |
| `openssh-client`    |`openssh-clients`                    |
| `ripgrep`           |`<gh releases>`                      |
| `rsync`             |`rsync`                              |
| `scp`               |`openssh-clients`                    |
| `screen`            |`NOT AVAILABLE`                      |
| `sed`               |`sed`                                |
| `socat`             |`socat`                              |
| `sudo`              |`sudo`                               |
| `ss`                |`NOT AVAILABLE`                      |
| `ssl-cert`          |`NOT AVAILABLE`                      |
| `tail`              |`<built in>`                         |
| `tar`               |`tar`                                |
| `time`              |`time`                               |
| `tldr`              |`NOT AVAILABLE (fedora only)`        |
| `tmux`              |`NOT AVAILABLE (fedora only)`        |
| `vim`               |`vim`                                |
| `wget`              |`wget`                               |
| `zip`               |`zip`                                |
| `zsh`               |`NOT AVAILABLE (fedora only)`        |
| **TOTAL SIZE**      | **411MB** (143MB compressed)        |

**languages runtimes and cloud tools**

| Tool or language    | ubi8 based image                    |
|---------------------|-------------------------------------|
|--------JAVA---------|-------------------------------------|
| `sdk`               |`<https://get.sdkman.io>`            |
| `java`              |`<11.0.12-open via sdkman>`          |
| `maven`             |`<via sdkman>`                       |
| `gradle`            |`<via sdkman>`                       |
|-------NODEJS--------|-------------------------------------|
| `nodejs`            |`nodejs`                             |
| `npm`               |`npm`                                |
| `yarn`              |`<via npm>`                          |
|--------GO-----------|-------------------------------------|
| `go`                |`go-toolset`                         |
|------PYTHON---------|-------------------------------------|
| `python`            |`python39`                           |
| `setuptools`        |`python39-setuptools`                |
| `pip`               |`python39-pip`                       |
|------CLOUD----------|-------------------------------------|
| `podman`            |`container-tools:rhel8`              |
| `buildah`           |`container-tools:rhel8`              |
| `skopeo`            |`container-tools:rhel8`              |
| `kubectl`           |`<kubernetes dnf repo>`              |
| `krew`              |`<gh releases>`                      |
| `helm`              |`<get.helm.sh>`                      |
| `kustomize`         |`<gh releases>`                      |
| `tkn`               |`<gh releases>`                      |
| `kn`                |`<gh releases>`                      |
| `terraform`         |`<releases.hashicorp.com>`           |
| `docker`            |`<download.docker.com>`              |
| `docker-compose`    |`<gh releases>`                      |
| **TOTAL SIZE**      | **2.72GB** (1.1GB compressed)       |

## Additional context

- [OpenShift WebTerminal Tooling](https://github.com/redhat-developer/web-terminal-tooling)
- [GitHub actions ubuntu virtual environment](https://github.com/actions/virtual-environments/blob/main/images/linux/Ubuntu2004-README.md) 
- GitPod [base](https://github.com/gitpod-io/workspace-images/blob/master/base/Dockerfile) and [full](https://github.com/gitpod-io/workspace-images/blob/master/full/Dockerfile) workspaces images

**Comment by ErnstHaagsman** [comment]
Could we add netstat and ss?

**Comment by ErnstHaagsman** [comment]
What do we think about man pages? Most Docker images exclude them, but in an interactive session [they're pretty useful](https://imgs.xkcd.com/comics/tar.png).

**Comment by l0rd** [comment]
@AlexKhotian @ErnstHaagsman I have included your suggestions to the list (including `jq` as a base terminal tool rather than a cloud tool). 

Now the thing is that I will start doing some tests because don't want the image size to go beyond, let's say, 4GB? For the Red Hat image we will use ubi-minimal. For the vendor neutral we can use debian as base.

**Comment by ErnstHaagsman** [comment]
I've been asking around a bit more. So far we're also missing: tmux, wget, openssh-client. What do we think about tar?

> Now the thing is that I will start doing some tests because don't want the image size to go beyond, let's say, 4GB? For the Red Hat image we will use ubi-minimal. For the vendor neutral we can use debian as base.

Looking forward to seeing the numbers!

**Comment by AlexKhotian** [comment]
@l0rd Would be great to optimize it for the size as well but without any critical cuts for the CX. Feel free to share a test version of the dockerfile so we can play around as well.

**Comment by ErnstHaagsman** [comment]
Got some more: `sed`, `scp` (included in `openssh-client`?), `screen`.

Others that were mentioned, but I'm personally not sure about: `rsync`, `netcat`, `socat`, `tldr`, `sdkman`, `jenv`.

**Comment by nickboldt** [comment]
I'd also suggest adding `skopeo` and `yq` (the python wrapper for jq, not the other one), and :-1: for `wget` (it's less secure than curl and has a smaller feature set).

**Comment by ArvinB** [comment]
Having only one dockerfile is preferred where `ubi` is a base minimum.

My two cents. I don't see these tools mentioned...

python (base tool)
opm (cloud tool)

**Comment by nickboldt** [comment]
How many versions of Java do we plan to include in this UBI-UDI image? If more than one, see https://github.com/eclipse/che/issues/16851 which might be a prereq for this task.

**Comment by justinmk3** [comment]
`fish`, `wget`, and `tldr` are very redundant with `bash`/`zsh`, `curl`, and `man`.

`rsync` is a great compliment to `ssh`, and enables powerful automation opportunities.

`tmux` would be really nice. I would even suggest dropping `screen`, unless it's a POSIX requirement...?

**Comment by l0rd** [comment]
I have updated the list including most of the tools mentioned in the comments. I have built 2 images based on the list:

- A base developer image (143MB compressed) with the base tools
- A universal developer image (1.1GB compressed) with languages and cloud tools

I pushed the Dockerfiles and other details in this new github repo: https://github.com/devfile/developer-images

**Comment by kadel** [comment]
> Devfile is a specification for containerized development environments. Such environments need to have a minimal set of command line developers tools.

Yes, but also devfile is highly customized for a specific project or type of project (language/framework).

If I have a Java project, why would I need python and nodejs in my developer image? I'm never going to use them with my Java project.

Another problem that I can see is that the list of the tools that you would like to include in the image can infinitely grow when you try to cover different versions of Java,Go,NodeJS.

**Comment by nickboldt** [comment]
well, it's either an infinite number of small containers to manage/maintain/update, or it's one all-in-one beast of a container that does everything (but chews up a lot of disk). 

This is the back-and-forth see-saw debate that has gone on for 30+ years in software. "Thin" vs. "thick", "server" vs. "client", "on prem" vs. "public cloud" ... you'll never make everyone happy, but you can swing the pendulum back and forth every few years and change the approach ... and the buzzwords... :D

**Comment by justinmk3** [comment]
> either an infinite number of small containers to manage/maintain/update, or it's one all-in-one beast of a container that does everything (but chews up a lot of disk).

3rd option: capture the common (80%) case with a container that's maybe 2x-4x the size of a minimal container (but 100x smaller than the maximum) but has the most common tools ("java latest LTS", "node latest LTS", ...).

**Comment by l0rd** [comment]
The Universal Developer Image [has been implemented](https://github.com/devfile/developer-images/) and is used by Eclipse Che and OpenShift Dev Spaces for every samples and as the default image when no devfile is found. We are happy with that and we got positives feedbacks from existing and new users.

I don't think we are going to use UDI in the devfile registry though as that is controversial so I am closing this issue as done.
